### PR TITLE
Fix unused_parens false positive when using binary operations

### LINKED ISSUE #71290: unused_parens false positive on addition of blocks?
In this example:

```rust
pub fn foo(a: bool, b: bool) -> u8 {
    (if a { 1 } else { 0 } + if b { 1 } else { 0 })
}
```

rustc throws a warning:
```
warning: unnecessary parentheses around block return value
 --> src/lib.rs:2:5
  |
2 |     (if a { 1 } else { 0 } + if b { 1 } else { 0 })
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: remove these parentheses
  |
  = note: `#[warn(unused_parens)]` on by default
```

Removing the parentheses as the message directs causes errors:

```
error: expected expression, found `+`
 --> src/lib.rs:2:27
  |
2 |     if a { 1 } else { 0 } + if b { 1 } else { 0 }
  |                           ^ expected expression

error[E0308]: mismatched types
 --> src/lib.rs:2:12
  |
2 |     if a { 1 } else { 0 } + if b { 1 } else { 0 }
  |     -------^-------------- help: consider using a semicolon here
  |     |      |
  |     |      expected `()`, found integer
  |     expected this to be `()`

error[E0308]: mismatched types
 --> src/lib.rs:2:23
  |
2 |     if a { 1 } else { 0 } + if b { 1 } else { 0 }
  |     ------------------^--- help: consider using a semicolon here
  |     |                 |
  |     |                 expected `()`, found integer
  |     expected this to be `()`

error: aborting due to 3 previous errors
```

It seems the warning is wrong, unless this is expected to compile

### Meta

`rustc --version --verbose`:
```
rustc 1.42.0 (b8cedc004 2020-03-09)
binary: rustc
commit-hash: b8cedc00407a4c56a3bda1ed605c6fc166655447
commit-date: 2020-03-09
host: x86_64-unknown-linux-gnu
release: 1.42.0
LLVM version: 9.0
```

This also happens on nightly.
```
rustc 1.44.0-nightly (ce93331e2 2020-04-17)
binary: rustc
commit-hash: ce93331e2cf21ac4b72a53854b105955919114e7
commit-date: 2020-04-17
host: x86_64-unknown-linux-gnu
release: 1.44.0-nightly
LLVM version: 9.0
```

This issue has been assigned to @mibac138 via [this comment](https://github.com/rust-lang/rust/issues/71290#issuecomment-623767470).

### PR DESCRIPTION
**Pull request by mibac138**
Fixes #71290 

r? @cuviper who provided instructions

### CODE REVIEW

File: src/test/ui/lint/issue-71290-unused-paren-binop.rs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+// check-pass
+// Make sure unused parens lint doesn't emit a false positive.
+// See https://github.com/rust-lang/rust/issues/71290 for details.
```
Comment (lcnr) [inline src/test/ui/lint/issue-71290-unused-paren-binop.rs:3]:
```suggestion
// See https://github.com/rust-lang/rust/issues/71290 for details.
#![deny(unused_parens)]
```
`ui/lint` tests set all lints to `allow` afaik, so this test already passes on the current beta as is.

Comment (mibac138) [reply src/test/ui/lint/issue-71290-unused-paren-binop.rs:3]: Thanks :)

**Review by lcnr — APPROVED** [review]

**Comment by cuviper** [comment]
I think this is going to need some level of recursion -- for example, this still fails:

```
fn z() -> u8 {
    ({ 0 } + 1 + 2)
}
```

In the AST, that will associate like `( ({ 0 } + 1) + 2 )`, so the LHS of the outer binop is another binop, which doesn't pass the `expr_requires_semi_to_be_stmt` test.

**Comment by mibac138** [comment]
I have now added recursion support and a relevant test case. I assume `Expr` can't create a cycle so this should be fine. Should I add an upper iteration bound anyway?

**Comment by cuviper** [comment]
> I assume `Expr` can't create a cycle so this should be fine. Should I add an upper iteration bound anyway?

I think it's fine -- for example, `contains_exterior_struct_lit` uses plain recursion on binary `lhs`/`rhs`.

bors r+

**Pull request merged**